### PR TITLE
OSSL_PARAM futures

### PR DESCRIPTION
**Pull request by paulidale**
Third alternative to #20888 and #20935.  Original report (& proposal) in #20812.

The intention here is to custom produce OSSL_PARAM parsers on a per algorithm basis.  This will easily be fastest of the three options.  It will be more branch predictor friendly, use less data space, ....

The only thing to look at is the [last commit](https://github.com/openssl/openssl/pull/20940/files#diff-505c0ffcec12976da4984af23bac9fa8d9961381e81cc1e97fea0c17c878dd2a) which outlines the proposed syntax.

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: doc/designs/fast-param-find.md
Comment (levitte) [inline doc/designs/fast-param-find.md:27]: We had this for the multiple CRT params in RSA keys, but we changed that to have indexed names instead.  I can't remember any other case where we did this.

Comment (t8m) [reply doc/designs/fast-param-find.md:27]: We concatenate the KDF INFO params.

Comment (t8m) [reply doc/designs/fast-param-find.md:27]: But maybe we could just say that you need to call the set param multiple times if you want this.

Comment (paulidale) [reply doc/designs/fast-param-find.md:27]: I'm not worried about the concatenation problem without calling set param multiple times.
I'm not even sure we actually need the _multiple_ flag.

Comment (paulidale) [reply doc/designs/fast-param-find.md:27]: More important is the general outline of the suggestion not the implementation specifics.
No rabbit holes.

File: doc/designs/fast-param-find.md
Comment (hlandau) [inline doc/designs/fast-param-find.md:70]: Yeah. I was thinking the codegen could '#define' a 'NUM_INDICES' or similar for the handwritten code to use to declare the array on the stack (or alternatively an array typedef), allowing the right length for each implementation. But if the entire function is being generated by codegen that's not needed. :+1:

File: doc/designs/fast-param-find.md
Comment (hlandau) [inline doc/designs/fast-param-find.md:41]: Is the idea that OSSL_PARAM_locate would be auto-substituted here? I think I'd probably have some codegen-specific symbol, maybe `@locate` or something.

**Review by hlandau — COMMENTED** [review]
Ah right, this is touching on the comments I just left in https://github.com/openssl/openssl/pull/20935#issuecomment-1544034240

I like the ideas outlined here. Codegen for uniform and monotonous aspects of code like param handling will reduce risk of bugs and allow greater performance than handwritten code. :+1:

File: doc/designs/fast-param-find.md
Comment (paulidale) [reply doc/designs/fast-param-find.md:41]: An oversight on my part.  We can't replace locate calls IMO.  They are still useful for repeated params -- but not a performance problem since they always come in sequentially (currently) and a strcmp is minimal effort.

**Comment by paulidale** [comment]
This approach has a number of drawbacks:

- I don't think we can use gperf to generate the hash tables meaning we'll need to implement our own
  - gpref does some odd things and doesn't return quite what we want
  - I doubt it's available everywhere
- Refactoring the cipher and digest code will be painful
  - The bulk of the parameter handling is buried in deeply intertwined macros
  - This might not be a major problem, it's only AEAD ciphers that aren't performant
- It will make for a lot of automatically generated files

File: doc/designs/fast-param-find.md
Comment (slontis) [inline doc/designs/fast-param-find.md:115]: Would a range check suffice instead of a full 256 byte table?

Comment (paulidale) [reply doc/designs/fast-param-find.md:115]: It would be slower but would certainly be possible.  gperf has an option to output 128 byte tables & we could reduce the character set further to alphanumerics, _ and -.

One thing to be cognisant of is that, in general, we will always be hashing strings that do match and that mismatches are very very much the exception.

Comment (paulidale) [reply doc/designs/fast-param-find.md:115]: If I thought it was future safe, I'd omit the strcmp entirely.
Unfortunately it isn't.

**Comment by levitte** [comment]
> * I don't think we can use gperf to generate the hash tables meaning we'll need to implement our own
>   * gpref does some odd things and doesn't return quite what we want
>   * I doubt it's available everywhere

Most of this (apart from odd things, which probably depends in the options used when generating) can be resolved by generating with `make update` and versioning the result instead of generating at build time.

We don't need to worry about output copyright, as stated in the [gperf manual, 3.5 The Copyright of the Output](https://www.gnu.org/software/gperf/manual/gperf.html#Output-Copyright)

**Comment by paulidale** [comment]
There is thing that's annoying, most of the rest can be optioned away.  gperf is actually doing more than we want.  We want to convert a string to an index.  Instead, gperf converts a string to the hashed item (by default the string) and a structure needs to be introduced to go back to the index.  It does this conversion using an index which would be ideal for our use.

Looking around and it seems like there are plenty of implementations of the table construction algorithm and it doesn't look too tricky, so it oughtn't to be considered an impediment at this stage.

**Comment by paulidale** [comment]
Looking for comments about this proposal

**Comment by beldmit** [comment]
What I am concerned about is downstream patch management. Red Hat has implemented a lot of FIPS specific changes in parameters validation, so I wonder how much efforts it would take to maintain them.

So probably `make update`-style will be mor downstream-friendly.

**Comment by t8m** [comment]
@beldmit I do not understand what would be the difference. Yes, it would take an initial effort to convert your added checks into the format @paulidale proposes, but I do not see why the further maintenance would be any harder to maintenance of the existing C sources of get/set params calls in providers.

It could be even easier to maintain the additions in this format.

**Comment by beldmit** [comment]
Generally not match - but, e.g. embracing the `case` sub-statement with parenthesis (to introduce local variables) will require some efforts, right?

**Comment by paulidale** [comment]
I agree with @t8m here, if there is downstream churn, it will be a once off.  I don't expect conversion of everything to happen in one go either.  The old way of searching and using params will remain viable of course.

The case sub-statements could always get {} around them to allow variable declaration.  I kind of hinted at this in the example but didn't mandate it.  There is also a prelude per function where variables can be more globally defined.  This generated code is still fairly ill-defined as to the specifics at this stage.

One thing I do see improving significantly is being able to get things which currently can only be set.  It's all pretty ad-hoc at the moment but mostly you cannot get the values you can set.  By automating code generation, the getters become a free bonus.

File: doc/designs/fast-param-find.md
Comment (hlandau) [inline doc/designs/fast-param-find.md:165]: nit: {

File: doc/designs/fast-param-find.md
Comment (hlandau) [inline doc/designs/fast-param-find.md:166]: nit: {

File: doc/designs/fast-param-find.md
Comment (hlandau) [inline doc/designs/fast-param-find.md:260]: nit: {

File: doc/designs/fast-param-find.md
Comment (hlandau) [inline doc/designs/fast-param-find.md:261]: nit: {

**Review by hlandau — CHANGES_REQUESTED** [review]
Looks good! Minor nits but otherwise LGTM.

File: doc/designs/fast-param-find.md
Comment (paulidale) [reply doc/designs/fast-param-find.md:165]: Doesn't need the { with a single statement body 😀

File: doc/designs/fast-param-find.md
Comment (paulidale) [reply doc/designs/fast-param-find.md:166]: Yeah, needs it here.

File: doc/designs/fast-param-find.md
Comment (paulidale) [reply doc/designs/fast-param-find.md:261]: and here.

Still prototype boilerplate code is meant to have errors

File: doc/designs/fast-param-find.md
Comment (levitte) [inline doc/designs/fast-param-find.md:135]: I would like to see the actual example.

**Review by levitte — CHANGES_REQUESTED** [review]

File: doc/designs/fast-param-find.md
Comment (levitte) [reply doc/designs/fast-param-find.md:135]: Ah, never mind, it's above.

**Comment by beldmit** [comment]
For the record: https://github.com/openssl/openssl/pull/20812 contains links to the original issue(s). Benchmark should cover some those scenarios.

File: doc/designs/fast-param-find.md
Comment (paulidale) [reply doc/designs/fast-param-find.md:135]: That is the example, output C code is further down (although unlikely to be all that close to the final output).

Until time is invested into implementing this, there won't & cannot be an actual example.

File: doc/designs/fast-param-find.md
Comment (paulidale) [reply doc/designs/fast-param-find.md:104]: Already catered for :wink:

Specifically in this text:
- "code" set to "no" skips code generation for this parameter, it defaults to "yes" which generates handlers.  This is useful when a parameter is duplicated with differenting types (e.g. utf8 string and integer).

**Comment by t8m** [comment]
IMO it is too early to merge the design document if we do not plan to implement this feature/refactoring in 3.2.

File: doc/designs/fast-param-find.md
Comment (levitte) [reply doc/designs/fast-param-find.md:104]: Sooooo.....  does that mean that those cases will have to be done entirely manually?
Might I suggest another approach, with something like `type => [ 'string', 'int' ]`, and have the code generator cope with that?

For perl branching, if the `type` value is an array, `ref($whatever->[type})` will return `"ARRAY"`, and if it's just a scalar (like a string), it will return `""`

Comment (paulidale) [reply doc/designs/fast-param-find.md:104]: Good suggestion.

**Comment by paulidale** [comment]
Okay, closing as done.

**Comment by t8m** [comment]
I'd keep the PR open. I do not see an issue if it is just lingering open until we find time to actually apply it.

**Comment by paulidale** [comment]
I don't think it's too early to merge.  Not merging designs is a long standing ongoing problem.
Having the design included but not implemented in 3.2 isn't the slightest issue.

**Comment by paulidale** [comment]
@openssl/otc @openssl/committers anyone have thoughts on merge now versus later?

**Review by hlandau — APPROVED** [review]
I have no objection to merging this now.

**Comment by paulidale** [comment]
It needs a second review as a formality.

File: doc/designs/fast-param-find.md
Comment (slontis) [inline doc/designs/fast-param-find.md:102]: attributes

**Review by slontis — APPROVED** [review]

**Comment by slontis** [comment]
doc-nits is complaining..

File: doc/designs/fast-param-find.md
Comment (paulidale) [reply doc/designs/fast-param-find.md:102]: fixed.

**Comment by paulidale** [comment]
Fixed the typo and addressed doc-nits -- everything was trivial so not necessary to reapprove IMO.

**Comment by paulidale** [comment]
CI's are happy.  I don't think this need reapproval from hlandau.

**Review by slontis — APPROVED** [review]

File: doc/designs/fast-param-find.md
Comment (paulidale) [reply doc/designs/fast-param-find.md:115]: Thinking more about this and I suspect it runs into code generation issues.  Processing for int and string would be quite different, so there would have to be some demarcation between the two so separate code could be generated.  I'm not seeing how this would be less invasive than the code = no approach I was thinking of.

Still, good to keep it in mind as an option.

**Comment by paulidale** [comment]
Proposal merged.

**Pull request closed without merging**
